File: inventory-service/src/test/java/com/ecommerce/inventoryservice/AbstractIntegrationTest.java
package com.ecommerce.inventoryservice;

import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.testcontainers.service.connection.ServiceConnection;
import org.testcontainers.mysql.MySQLContainer;

@SpringBootTest
public abstract class AbstractIntegrationTest {

    // @Testcontainers를 사용하면 테스트 클래스가 끝날 때 static 컨테이너도 종료됨
    // 여러 테스트 클래스가 공유할 수 있도록 static 초기화 블록으로 직접 시작
    // JVM 종료 시 Ryuk이 컨테이너를 자동으로 정리함
    @ServiceConnection
    static final MySQLContainer MYSQL = new MySQLContainer("mysql:8.0");

    static {
        MYSQL.start();
    }
}

File: review-service/src/test/kotlin/com/ecommerce/reviewservice/AbstractIntegrationTest.kt
package com.ecommerce.reviewservice

import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.testcontainers.service.connection.ServiceConnection
import org.testcontainers.mysql.MySQLContainer

@SpringBootTest
abstract class AbstractIntegrationTest {

    companion object {
        // @Testcontainers 대신 static 초기화로 JVM 전체에서 컨테이너 공유
        // Ryuk이 JVM 종료 시 자동 정리
        // @JvmField로 실제 static 필드가 되어야 @ServiceConnection이 적용됨
        @ServiceConnection
        @JvmField
        val MYSQL: MySQLContainer = MySQLContainer("mysql:8.0")

        init {
            MYSQL.start()
        }
    }
}
